### DISCUSSION TITLE
Not notified of the break in SRTP connection

### DISCUSSION THREAD

**Discussion by hills** [baresip/baresip#2310] [General]
When using baresip.h as an API, I don't seem to have a way to be notified of the break in SRTP connection.

The software recognises it from the peer, as seen in the console, but there is no action follows:

```
dtls_srtp: dtls-connection closed (Connection reset by peer)
```

Intuitively the test on `comp->negotiated` seems to be inverted. Is this a bug? Or merely that I do not understand the intent behind this code.

It seems likely the intent is to propagate errors to the caller after dtls_estab_handler() has taken place, not before., as the latter makes its own calls to the errorh. I'm happy to be corrected, I just need the appropriate way of detecting a closed connection. Many thanks.

```
--- a/modules/dtls_srtp/dtls_srtp.c
+++ b/modules/dtls_srtp/dtls_srtp.c
@@ -266,9 +266,9 @@ static void dtls_close_handler(int err, void *arg)
        info("dtls_srtp: dtls-connection closed (%m)\n", err);

        comp->tls_conn = mem_deref(comp->tls_conn);

-       if (!comp->negotiated) {
+       if (comp->negotiated) {

                if (comp->ds->sess->errorh)
                        comp->ds->sess->errorh(err, comp->ds->sess->arg);
        }
```

**Comment by alfredh** [comment]
The cipher material from the DTLS connection is used to generate keys for SRTP.

This only happens once, right after the DTLS connection is established.
After that it is not used anymore.

If the close handler is called before the established handler, it means that something failed.

A different approach would be to close the DTLS connection
in the established handler, right after SRTP keys are generated.
This has to be interop tested though...

**Comment by hills** [comment]
Thanks. That begins to clarify, I think. Please correct me in any errors here:

You're saying that any 'errors' are simply not relevant after the negotiation. Interesting.

I was seeing the "dtls-connection closed" message only on shutting down the peer (in this case, closing the Chromium browser). It does not happen any earlier. That led me to believe this was end of the DTLS/SRTP session, rather than just end of the DTLS negotiation.

So therefore the browser could legally issue this close at any time, it just happens to do so on closing the window.

Therefore, what is the correct way to detect the 'close' of the SRTP session if, for example the peer goes away?

Thanks

**Reply by alfredh** [reply]
to detect livelyness of the peer, you have to send periodic STUN Binding requests.
If the peer does not answer, you can close the session

**Reply by hills** [reply]
Thanks @alfredh. I'm returning to take a look at this today.

I'm not using STUN API directly (and no STUN or TURN server as we have full connectivity), but I am using baresip "ice" module by calling it in code and passing in the SDP.

With a breakpoint on stun_request() I can see in practice that this only called at the start, not periodically.

But I don't see a path through the baresip "mnat" API to send periodically. How does baresip do this or is it expected that it just keeps sending if the peer goes away?

I'm (very) slowly gaining more understanding of how the ICE and STUN modules in baresip & re fit together. Clearly I need to understand more of STUN beyond discovering IP addresses. Perhaps the "conncheck" process is where I need to be looking. Any pointers to help my understanding would be appreciated. I assume it isn't the responsibilty of RTP or RTCP to detect the peer going away then. Many thanks.

**Reply by hills** [reply]
I spent more time, but having difficult to see a path to do this.

My server using baresip is ICE_ROLL_CONTROLLED (because the client web browser sends initial offer). It sends only STUN indicator instead of STUN request.

This is different to browser-to-browser, where both sides send STUN request like you suggest.

I read a lot of code to try and understand the protocols and the structure. I don't think periodic STUN binding happens even for ICE_ROLL_CONTROLLED or ICE_ROLL_CONTROLLING? Always STUN indication.

I think maybe I missed something. This post shows the lack of clarity in my thinking. @alfredh, do you have any suggestions please? I looked at many approaches, all seem bad:

* Forcing the server into ICE_ROLL_CONTROLLING. Seems wrong, and still code is missing to make STUN requests and report disconnection back from mnat
* Base it on existing 'connchk' process, as this is the only place which already has any kind of loop to `stun_request`
* Somewhere I could insert calls to libre `stun_keepalive_enable` which is currenly unused (except elsewhere by baresip "stun" module)
* In ```icem_comp_keepalive``` change from STUN indicator to STUN request, but still need a path back up through mnat to report disconnection.
* Implement periodic STUN binding in my own code (maybe use `stun_keepalive_enable`), but it seems I'm several layers from knowing the correct socket and, importantly, when I can start polling.
* Don't change the messages, instead just use the existing client's STUN requests. This assumes the client always makes STUN request. Create some path for a callback from `libre/src/ice/stunsrv.c` to say the connection is broken.

Any advice appreciated, many thanks.
